lib/class-wp-rest-widget-updater-controller.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Start: Include for phase 2
+ * Widget Updater REST API: WP_REST_Widget_Updater_Controller class
+ *
+ * @package gutenberg
+ * @since 5.2.0
+ */
+
+/**
+ * Controller which provides REST endpoint for updating a widget.
+ *
+ * @since 5.2.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Widget_Updater_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'widgets';
+	}
+
+	/**
+	 * Registers the necessary REST API route.
+	 *
+	 * @access public
+	 */
+	public function register_routes() {
+			register_rest_route(
+				$this->namespace,
+				// Regex representing a PHP class extracted from http://php.net/manual/en/language.oop5.basic.php.
+				'/' . $this->rest_base . '/(?P<identifier>[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*)/',
+				array(
+					'args' => array(
+						'identifier' => array(
+							'description' => __( 'Class name of the widget.', 'gutenberg' ),
+							'type'        => 'string',
+						),
+					),
+					array(
+						'methods'             => WP_REST_Server::EDITABLE,
+						'permission_callback' => array( $this, 'compute_new_widget_permissions_check' ),
+						'callback'            => array( $this, 'compute_new_widget' ),
+					),
+				)
+			);
+	}
+
+	/**
+	 * Checks if the user has permissions to make the request.
+	 *
+	 * @since 5.2.0
+	 * @access public
+	 *
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function compute_new_widget_permissions_check() {
+		// Verify if the current user has edit_theme_options capability.
+		// This capability is required to access the widgets screen.
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'widgets_cannot_access',
+				__( 'Sorry, you are not allowed to access widgets on this site.', 'gutenberg' ),
+				array(
+					'status' => rest_authorization_required_code(),
+				)
+			);
+		}
+		return true;
+	}
+
+	/**
+	 * Returns the new widget instance and the form that represents it.
+	 *
+	 * @since 5.2.0
+	 * @access public
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function compute_new_widget( $request ) {
+		$url_params = $request->get_url_params();
+
+		$widget = $request->get_param( 'identifier' );
```

Comment:
We can't use the class name (e.g. `'WP_Widget_Search'`) to identify which widget to render, as one can register a widget in WordPress without using `WP_Widget`:

```php
function my_widget_callback() {
	return '<marquee>This is a legacy widget! It doesn't support options.</marquee>';
}
wp_register_sidebar_widget( 'my-widget', 'My cool widget', 'my_widget_callback' );
```

Widgets registered this way do not support user customisable settings and they cannot be used more than once per sidebar.

I haven't looked enough at your PR to make a recommendation on what to do, but I am thinking you might need to use the _widget ID_ (e.g. `'search-2'`) instead of the class name (e.g. `'WP_Widget_Search'`) to reference which widget to render. The widget can then be looked up by its ID via the global `$wp_registered_widgets` array.

I'm currently dealing with the same roadblock while working on how to store blocks within our widgets architecture, so let me know if you want to chat about widgets—it's confusing! 🙂

Reply:
OK, so I am thinking that we may have to make the block contain only the widget's identifier as an attribute, and nothing else:

```
<!-- wp:legacy-widget { "id": "search-2" } /-->
```

Rendering the widget's frontend would work by calling `$wp_registered_widgets[ $id ]['callback']`. Both legacy widgets and `WP_Block` widgets support this. Check out the source of `dynamic_sidebar()` in `wp-includes/widgets.php` for an example of this.

Rendering the widget's backend would work by calling `$wp_registered_widget_controls[ $id ]['callback']`. This will output a `<form>` which can be returned to the client. Both legacy widgets and `WP_Block` widgets support this. Check out the source of `wp_widget_control()` in `wp-admin/includes/widgets.php` for an example of this.

Updating the widget would work similar to how we currently support meta boxes: on save, we serialise the widget's `<form>` and send it off to the server. The server then calls `$wp_registered_widget_controls[ $id ]['callback']` which will take care of inspecting `$_POST` and making updates as necessary. Both legacy widgets and `WP_Block` widgets support this. Check out the source of `wp_ajax_save_widget()` in `wp-admin/includes/ajax-actions.php` for an example of this.

Does that make sense, or have I missed something? I'm still learning how widgets are implemented in WordPress so take everything I say with a grain of salt! 🧂

Reply:
> We can't use the class name (e.g. 'WP_Widget_Search') to identify which widget to render, as one can register a widget in WordPress without using WP_Widget:

function my_widget_callback() {
	echo '<marquee>This is a legacy widget!</marquee>';
}
wp_register_sidebar_widget( 'my-widget', 'My cool widget', 'my_widget_callback' );
Widgets registered this way do not support user customisable settings have to manage their settings manually, and they cannot be used more than once per sidebar.


Hi @noisysocks, good catch! I missed this type of widgets. I updated the code to correct the problem.

The sample code you provided now works as expected:

![mar-03-2019 12-52-14](https://user-images.githubusercontent.com/11271197/53695397-422a0080-3db3-11e9-8d17-3ca7a8c04ea6.gif)

Reply:
> OK, so I am thinking that we may have to make the block contain only the widget's identifier as an attribute, and nothing else:

```
<!-- wp:legacy-widget { "id": "search-2" } /-->
```
Hi @noisysocks 
I thought about this option. It would probably have been easier to create an initial version following this path.

This path has some problems though:
- The source of truth is not only the post content we are referencing something in the options database. It may not be clear that changing something in the options database (e.g: clean it) will change the way a normal post with a very simple legacy widget looks.

- When we copy the code of the widget ```<!-- wp:legacy-widget { "id": "search-2" } /-->```, and past it somewhere the expectation is that changing the pasted block will not change the original block. If that behaviour is what the user desires, the user should use a legacy block.
- The user may want to use a new widget right after installing it, for which no widget instance exists on the database. So we would need a mechanism to create new widget instances (a side effect of using the block the block).
- If the user undoes a legacy widget usage immediately after a new instance was created, we need an undo side effect removing the widget from the database (undo side effects seems like something that may cause troubles). Redo side effect would also be needed to create a widget instance again.
- If the user removes a legacy widget instance, we would have some troubles. We would need to decide if we remove it from the database or not. If we remove it from the database, we are in trouble if it was referenced meanwhile in other post. If we don't remove it from the database, each time legacy widgets are inserted the number of widget instances grows. Growing the number of widget instances, may not scale well given the way widgets are stored (serialized PHP objects).

Saving the widget instance as standard block attributes seemed to overcome this limitations.